src/gt4py/next/utils.py

```diff
@@ -285,213 +282,100 @@ def equalize_tuple_structure(
-    ]
-    all_keywords = {
-        key for key, param in params.items() if param.kind != inspect.Parameter.POSITIONAL_ONLY
-    }
-    all_args_count = len(params)
-
-    if allow_kwargs_mutation and not sort_kwargs:
-        # If we can mutate kwargs and don't need to sort them,
-        # we can use the passed kwargs dict, which should be
-        # in the correct order after popping the positional args.
-        canonical_kwargs_expr = "kwargs"
-    else:
-        kwargs_items_exprs = [f"{name!r}: kwargs[{name!r}]" for name in kwonly_names]
-        canonical_kwargs_expr = f"{{ {str.join(', ', kwargs_items_exprs)} }}"
-
-    @functools.cache
-    def canonicalizer_factory(
-        passed_pos_args_count: int, passed_kwargs_keys: tuple[str, ...]
-    ) -> CallArgsCanonicalizer:
-        # This function and the generated canonicalizer are optimized for performance
-        if passed_pos_args_count > pos_args_count:
-            raise ValueError(
-                f"Too many positional arguments (expected {pos_args_count}, got {passed_pos_args_count})."
-            )
-        passed_kwargs_set = {*passed_kwargs_keys}
-        if unexpected_kwargs := (passed_kwargs_set - all_keywords):
-            raise ValueError(f"Got unexpected keyword arguments: {unexpected_kwargs}.")
-        if missing_kwargs := ({*kwonly_names} - passed_kwargs_set):
-            raise ValueError(f"Missing keyword arguments: {missing_kwargs}.")
-        if (total_arg_count := (passed_pos_args_count + len(passed_kwargs_keys))) > all_args_count:
-            raise ValueError(
-                f"Too many total arguments for the passed signature (expected {all_args_count}, got {total_arg_count})."
-            )
+    """
 
-        if pos_args_names & passed_kwargs_set:
-            unpack_args_stmt = (
-                (f"{str.join(', ', (f'a{i}' for i in range(passed_pos_args_count)))}, = args")
-                if passed_pos_args_count
-                else "# No args to unpack"
-            )
+    params = []
+    pos_args = []
+    key_args = []
+    var_pos_arg: str | None = None
+    var_key_arg: str | None = None
+
+    for key, param in signature.parameters.items():
+        params.append(param.replace(annotation=param.empty))  # remove annotations to avoid issues
+        match param.kind:
+            case inspect.Parameter.POSITIONAL_ONLY | inspect.Parameter.POSITIONAL_OR_KEYWORD:
+                pos_args.append(f"{key}")
+            case inspect.Parameter.VAR_POSITIONAL:
+                var_pos_arg = key
+                pos_args.append(f"*{key}")
+            case inspect.Parameter.KEYWORD_ONLY:
+                key_args.append(f"{key!r}: {key}")
+            case inspect.Parameter.VAR_KEYWORD:
+                var_key_arg = key
+                key_args.append(f"**{key}")
+
+    canonicalizer_signature = inspect.Signature(
+        parameters=params, return_annotation="tuple[tuple, dict[str, Any]]"
```

Comment:
What's the reason to annotate the return type? For looking at the `str`? It's not useful for tools, right?